core/vm/contracts_write.go

```diff
@@ -150,3 +169,168 @@ func (c *stakingPrecompile) RunWriteCapable(
 	//}
 	return nil, errors.New("[StakingPrecompile] Received incompatible stakeMsg from staking.ParseStakeMsg")
 }
+
+var abiCrossShardXfer abi.ABI
+
+func init() {
+	// msg.Value is used for transfer and is also a parameter
+	// otherwise it might be possible for a user to retrieve money from the precompile
+	// that was sent by someone else prior to the hard fork
+	// contract.Caller is used as fromAddress, not a parameter
+	// originating ShardID is pulled from the EVM object, not a parameter
+	crossShardXferABIJSON := `
+	[
+	  {
+	    "inputs": [
+		  {
+	        "internalType": "uint256",
+	        "name": "value",
+	        "type": "uint256"
+	      },
+		  {
+	        "internalType": "address",
+	        "name": "to",
+	        "type": "address"
+	      },
+		  {
+	        "internalType": "uint64",
+	        "name": "toShardID",
+	        "type": "uint32"
+	      }
+	    ],
+	    "name": "crossShardTransfer",
+	    "outputs": [],
+	    "stateMutability": "payable",
+	    "type": "function"
+	  }
+	]`
+	var err error
+	abiCrossShardXfer, err = abi.JSON(strings.NewReader(crossShardXferABIJSON))
+	if err != nil {
+		// means an error in the code
+		panic("Invalid cross shard transfer ABI JSON")
+	}
+}
+
+type crossShardXferPrecompile struct {
+	address common.Address
+}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *crossShardXferPrecompile) RequiredGas(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) (uint64, error) {
+	// we just charge the intrinsic gas here
+	// using data that includes: fromAddress, toAddress, fromShardID, toShardID, value
+	var payload []byte = make([]byte, 0)
```

Comment:
The gas charged on normal cross-shard transaction is just the gas for a normal token transfer. I think we can stick with it.

Reply:
The extra gas is being charged to prevent smart contracts from allowing EOAs to subsidize these transfers via `delegatecall`. The same was tested out in the staking precompile.
https://github.com/MaxMustermann2/harmony-staking-precompiles/blob/28c9f0761c3bd3cda5ee8dc9e42d4dd3ecd75e92/contracts/example/TrySubsidizeContract.sol#L29

A smart contract which delegate calls 0xF9 multiple times on behalf of an EOA will be able to save the EOA the base gas of 21000 per call and the system will just charge them for the overall data + 21000 only once.

Reply:
sounds good.